File: mobile/lib/features/cv/domain/entities/experience.dart
import 'copy_with_sentinel.dart';

/// Experience professionnelle d'un CV, entite de domaine immuable et pure.
final class Experience {
  final int? id;
  final String? entreprise;
  final String? poste;
  final String? lieu;
  final DateTime? dateDebut;
  final DateTime? dateFin;
  final String? description;
  final bool actuel;

  const Experience({
    this.id,
    this.entreprise,
    this.poste,
    this.lieu,
    this.dateDebut,
    this.dateFin,
    this.description,
    this.actuel = false,
  });

  Experience copyWith({
    Object? id = unsetSentinel,
    Object? entreprise = unsetSentinel,
    Object? poste = unsetSentinel,
    Object? lieu = unsetSentinel,
    Object? dateDebut = unsetSentinel,
    Object? dateFin = unsetSentinel,
    Object? description = unsetSentinel,
    bool? actuel,
  }) {
    return Experience(
      id: identical(id, unsetSentinel) ? this.id : id as int?,
      entreprise: identical(entreprise, unsetSentinel)
          ? this.entreprise
          : entreprise as String?,
      poste: identical(poste, unsetSentinel) ? this.poste : poste as String?,
      lieu: identical(lieu, unsetSentinel) ? this.lieu : lieu as String?,
      dateDebut: identical(dateDebut, unsetSentinel)
          ? this.dateDebut
          : dateDebut as DateTime?,
      dateFin: identical(dateFin, unsetSentinel)
          ? this.dateFin
          : dateFin as DateTime?,
      description: identical(description, unsetSentinel)
          ? this.description
          : description as String?,
      actuel: actuel ?? this.actuel,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is Experience &&
      other.id == id &&
      other.entreprise == entreprise &&
      other.poste == poste &&
      other.lieu == lieu &&
      other.dateDebut == dateDebut &&
      other.dateFin == dateFin &&
      other.description == description &&
      other.actuel == actuel;

  @override
  int get hashCode => Object.hash(
        id,
        entreprise,
        poste,
        lieu,
        dateDebut,
        dateFin,
        description,
        actuel,
      );
}

File: backend/src/main/java/com/cvmobile/cv/domain/model/Experience.java
package com.cvmobile.cv.domain.model;

import java.time.LocalDate;

/**
 * Experience professionnelle figurant sur un CV (valeur de domaine immuable).
 *
 * <p>Tous les champs sont facultatifs : un CV peut etre edite partiellement.
 * Le drapeau {@code actuel} vaut {@code false} par defaut lorsqu'il n'est pas
 * fourni, evitant un booleen null ambigu dans le reste du domaine.
 *
 * @param id          identifiant technique, {@code null} tant que non persiste
 * @param entreprise  employeur, ou {@code null}
 * @param poste       intitule du poste, ou {@code null}
 * @param lieu        localisation, ou {@code null}
 * @param dateDebut   debut de la periode, ou {@code null}
 * @param dateFin     fin de la periode, ou {@code null} si en cours
 * @param description missions et realisations, ou {@code null}
 * @param actuel      poste occupe actuellement (jamais {@code null})
 */
public record Experience(
        Long id,
        String entreprise,
        String poste,
        String lieu,
        LocalDate dateDebut,
        LocalDate dateFin,
        String description,
        boolean actuel) {

    /**
     * Construit une experience en normalisant les champs texte. {@code actuel}
     * est ramene a {@code false} lorsqu'il est {@code null}.
     */
    public static Experience of(
            Long id,
            String entreprise,
            String poste,
            String lieu,
            LocalDate dateDebut,
            LocalDate dateFin,
            String description,
            Boolean actuel) {
        return new Experience(
                id,
                DomainText.normalize(entreprise),
                DomainText.normalize(poste),
                DomainText.normalize(lieu),
                dateDebut,
                dateFin,
                DomainText.normalize(description),
                actuel != null && actuel);
    }

    /**
     * Derive une copie de cette experience avec une nouvelle description
     * (utilise par l'adaptation IA des variantes, sans muter l'original).
     */
    public Experience withDescription(String newDescription) {
        return new Experience(
                id, entreprise, poste, lieu, dateDebut, dateFin,
                DomainText.normalize(newDescription), actuel);
    }
}
